packages/@cdktf/provider-generator/lib/get/constructs-maker.ts

```diff
@@ -273,10 +351,137 @@ export class ConstructsMaker {
+      );
+    } catch (e) {
+      // ignore
+      logger.debug(`Failed to read versions file: ${e}`);
+    }
+
+    const versions = allowedConstraints.reduce((acc, constraint) => {
+      const provider = Object.entries(previousVersions).find(([name]) =>
+        // This could be more refined, but it's good enough for now
+        name.endsWith(constraint.fqn)
+      );
+
+      if (provider) {
+        const [name, version] = provider;
+        acc[name] = version;
+      }
+
+      return acc;
+    }, this.versions);
+
+    logger.debug(
+      `Writing versions file (${filePath}): ${JSON.stringify(
+        versions,
+        null,
+        2
+      )}`
+    );
+
+    this.code.openFile(filePath);
+    this.code.line(JSON.stringify(versions, null, 2));
+    this.code.closeFile(filePath);
+    return filePath;
+  }
+
+  public async removeFoldersThatShouldNotExist(
+    constraintsThatShouldExist: TerraformDependencyConstraint[]
+  ) {
+    logger.debug(
+      `Removing providers except for ${JSON.stringify(
+        constraintsThatShouldExist,
+        null,
+        2
+      )}`
+    );
+
+    // All languages besides TS keep their providers in the same folders as modules
+    // this makes it impossible for us to distinguish a no longer required provider
+    // from a manually written construct or a module
+    if (!this.isJavascriptTarget) {
+      return;
+    }
+
+    let filesInProviders: string[] = [];
+    // TODO: this path is different depending on the language
+    const providersFolder = path.resolve(this.codeMakerOutdir, "providers");
+    try {
+      filesInProviders = await fs.readdir(providersFolder);
+    } catch (e) {
+      logger.debug(
+        `Error listing files in providers folder '${providersFolder}': ${e}`
+      );
+    }
+
+    const folders = filesInProviders.filter((file) =>
+      fs
+        .statSync(path.resolve(this.codeMakerOutdir, "providers", file))
+        .isDirectory()
+    );
+
+    return folders.forEach((folder) => {
+      const shouldExist = constraintsThatShouldExist.some(
+        (constraint) => constraint.name === folder
+      );
+
+      if (!shouldExist) {
+        logger.debug(`Removing folder ${folder} from providers`);
+        fs.removeSync(path.resolve(this.codeMakerOutdir, "providers", folder));
+      }
+    });
+  }
+
+  // emits a constraints.json file with a map of the used provider fqpns and version constraints
+  // this is used for caching purposes
+  private emitConstraintsFile(
+    allowedConstraints: TerraformDependencyConstraint[]
+  ) {
+    const filePath = "constraints.json";
+
     this.code.openFile(filePath);
-    this.code.line(JSON.stringify(this.versions, null, 2));
+    this.code.line(
+      JSON.stringify(
+        allowedConstraints
+          .sort((a, b) => a.fqn.localeCompare(b.fqn))
+          .reduce(
+            (carry, item) => ({
+              ...carry,
+              [item.fqn]: item.version,
+            }),
+            {}
+          ),
```

Comment:
Object.fromEntries is not available in our TS target, we'd need to add the lib target as es2019.
Plus I find the reduce more readable since the logical pieces are closer to another. With your solution you would first wonder why you give an array of arrays back before looking to the Object.fromEntries on the outside